<html><head>
	<meta content="text/html; charset=utf-8" http-equiv="CONTENT-TYPE">
	<title></title>
	<meta content="LibreOffice 3.5  (Linux)" name="GENERATOR">
	<meta content="Koen " name="AUTHOR">
	<meta content="20120514;10115900" name="CREATED">
	<meta content="Koen " name="CHANGEDBY">
	<meta content="20120523;17082400" name="CHANGED">
	<style type="text/css">
	<!--
		@page { margin: 0.79in }
		p { margin-bottom: 0.08in }
		h1 { margin-bottom: 0.08in }
		h1.western { font-family: "Arial", sans-serif; font-size: 16pt }
		h1.cjk { font-family: "DejaVu Sans"; font-size: 16pt }
		h1.ctl { font-family: "Lohit Hindi"; font-size: 16pt }
		h2 { margin-bottom: 0.08in }
		h2.western { font-family: "Arial", sans-serif; font-size: 14pt; font-style: italic }
		h2.cjk { font-size: 14pt; font-style: italic }
		h2.ctl { font-family: "Lohit Hindi"; font-size: 14pt; font-style: italic }
		h3 { margin-bottom: 0.08in }
		h3.western { font-family: "Arial", sans-serif }
		h3.ctl { font-family: "Lohit Hindi" }
		h4 { margin-bottom: 0.08in }
		h4.western { font-family: "Arial", sans-serif; font-size: 11pt; font-style: italic }
		h4.cjk { font-size: 11pt; font-style: italic }
		h4.ctl { font-family: "Lohit Hindi"; font-size: 11pt; font-style: italic }
		a:link { so-language: zxx }
	-->
	</style>
</head>
<body lang="en-US" dir="LTR">
<h1 align="CENTER" class="western"></h1>
<h1 align="CENTER" class="western"></h1>
<h1 align="CENTER" class="western"><font size="7" style="font-size: 44pt"><font color="#0000ff">i</font>Regulon</font></h1>
<p align="CENTER"><br><br>
</p>
<h1 class="western"><br><br>
</h1>

<p style="margin-bottom: 0in">iRegulon was developed in the
Laboratory of Computational Biology (S. Aerts lab), University of
Leuven, Belgium.</p>
<p style="margin-bottom: 0in">This work is under writing process. For
the moment please cite the webpage (lab description) and the Methods
paper.</p>
<p style="margin-bottom: 0in">Site:</p>
<p style="margin-left: 0.49in; margin-bottom: 0in"><a href="http://med.kuleuven.be/cme/subpage.html?section=laboratories&amp;subsection=laboratory-of-computational-biology">http://med.kuleuven.be/cme/subpage.html?section=laboratories&amp;subsection=laboratory-of-computational-biology</a></p>
<p style="margin-bottom: 0in">Paper: 
</p>
<p style="margin-bottom: 0in">Potier, D., Atak, Z.K., Sanchez, M.N.,
Herrmann, C. and Aerts S. (2012). 
</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Methods in
Molecular Biology vol. 786,</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Chapter 18: Using
cisTargetX to Predict Transcriptional Targets and Networks in
Drosophila.</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Springer Science</p>
<p style="margin-bottom: 0in">Developers: Koen Herten &amp; Bram Van
de Sande.</p>
<p style="margin-bottom: 0in">Copyright 2012 KU Leuven.</p>
<p style="margin-bottom: 0in">Current version: iRegulon Beta v0.10.3</p>
<p style="margin-bottom: 0in">Contact: 	Bram Van de Sande
		<a href="mailto:Bram.VandeSande@med.kuleuven.be">Bram.VandeSande@med.kuleuven.be</a></p>
<p style="margin-bottom: 0in">		Stein Aerts
			<a href="mailto:Stein.Aerts@med.kuleuven.be">Stein.Aerts@med.kuleuven.be</a></p>
<div dir="LTR" id="Table of Contents1">
	<div dir="LTR" id="Table of Contents1_Head">
		<p style="margin-top: 0.17in; page-break-before: always; page-break-after: avoid">
		<font face="Arial, sans-serif"><font size="4" style="font-size: 16pt"><b>Table
		of Contents</b></font></font></p>
	</div>
	<p style="margin-bottom: 0in"><a href="#Tutorial">Tutorial</a></p>
	<p style="margin-left: 0.2in; margin-bottom: 0in"><a href="#Installation">Installation</a></p>
	<p style="margin-left: 0.2in; margin-bottom: 0in"><a href="#Summary">Summary</a></p>
	<p style="margin-bottom: 0in"><a href="#Methods paper">Methods paper</a></p>
	<p style="margin-bottom: 0in"><a href="#Site">Site</a></p>
	<p style="margin-bottom: 0in"><a href="#Current version">Current version</a></p>
	<p style="margin-left: 0in; margin-bottom: 0in"><a href="#Contact">Contact</a></p>
	<p style="margin-left: 0in; margin-bottom: 0in"><a href="#Case study">Case study </a></p>
	<p style="margin-left: 0.2in; margin-bottom: 0in"><a href="#Gene based">gene based</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Gene based Input">Input</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Gene based Running the analysis">Running the
	analysis</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Gene based Interpret the results">Interpret the
	results</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Gene based Visualization">Visualization</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Gene based Save and Export">Save and
	Export</a></p>
	<p style="margin-left: 0.2in; margin-bottom: 0in"><a href="#Region based">Region based</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Input">Input</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Running the analysis">Running the
	analysis</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Interpret the results">Interpret the
	results</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Visualization">Visualization</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Save and Export">Save and
	Export</a></p>
	<p style="margin-left: 0.4in; margin-bottom: 0in"><a href="#Region based Link Out">Link Out</a></p>
</div>
<h1 class="western"><br><br>
</h1>
<h1 style="page-break-before: always" class="western"><a name="Tutorial">Tutorial</a></h1>
<p style="margin-bottom: 0in"><br>
</p>
<p style="margin-bottom: 0in">For this tutorial the gene set
YY1_XiRen of the AMADEUS benchmark [<font  style="font-size: 11pt"><span lang="en-US">Linhart
C., Halperin, Y.,and  Shamir, R. (2008) Transcription factor and
microRNA motif discovery: The Amadeus platform and a compandium of
metazoan target sets, </span></font><font  style="font-size: 11pt"><span lang="en-US"><i>Genome
Research 18</i></span></font><font  style="font-size: 11pt"><span lang="en-US">,
1180-1189</span></font>] is used. <font  style="font-size: 11pt"><span lang="en-US">This
is a set of 698 genes. The known TF is YY1. This TF regulates histone
modifications (deacetylases and acetyltransferases). [GeneCards, </span></font><font  style="font-size: 11pt"><span lang="en-US"><span style="font-weight: normal">YY1
transcription factor [www], 2012, GeneCards,</span></span></font><font  style="font-size: 11pt"><span lang="en-US">
</span></font><a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=YY1&amp;search=yy1">http://www.genecards.org/cgi-bin/carddisp.pl?gene=YY1&amp;search=yy1</a><font  style="font-size: 11pt"><span lang="en-US">]</span></font></p>
<p style="margin-bottom: 0in"><br>
</p>
<h2 class="western"><font  style="font-size: 11pt"><a name="Installation">Installation</a></font></h2>
<p><font  style="font-size: 11pt">Place the jar in your
Cytoscape plugins folder. Cytoscape_v2.8.* &gt; plugins.</font></p>
<h2 class="western"><font  style="font-size: 11pt"><a name="Summary">Summary</a></font></h2>
<p><font  style="font-size: 11pt">iRegulon is a Cytoscape
plugin.</font></p>
<p><font  style="font-size: 11pt">iRegulon will predict
transcription factors (TFs), enhancers and target genes (TGs).</font></p>
<p style="margin-bottom: 0in">iRegulon was developed in the
Laboratory of Computational Biology (S. Aerts lab), University of
Leuven, Belgium.</p>
<p style="margin-bottom: 0in"></p>
<p style="margin-bottom: 0in">This work is under writing process. For
the moment please cite the webpage (lab description) and the Methods
paper.</p>
<p style="margin-bottom: 0in"></p>
<p style="margin-bottom: 0in"><a name="Site">Site:</a></p>
<p style="margin-left: 0.49in; margin-bottom: 0in"><a href="http://med.kuleuven.be/cme/subpage.html?section=laboratories&amp;subsection=laboratory-of-computational-biology">http://med.kuleuven.be/cme/subpage.html?section=laboratories&amp;subsection=laboratory-of-computational-biology</a></p>
<p style="margin-left: 0.49in; margin-bottom: 0in"></p>
<p style="margin-bottom: 0in"><a name="Methods paper">Paper:</a> 
</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Potier, D., Atak,
Z.K., Sanchez, M.N., Herrmann, C. and Aerts S. (2012). 
</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Methods in
Molecular Biology vol. 786,</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Chapter 18: Using
cisTargetX to Predict Transcriptional Targets and Networks in
Drosophila.</p>
<p style="margin-left: 0.49in; margin-bottom: 0in">Springer Science</p>
<p style="margin-bottom: 0in"></p>
<p style="margin-bottom: 0in">Developers: Koen Herten &amp; Bram Van
de Sande.</p>
<p style="margin-bottom: 0in">Copyright 2012 KU Leuven.</p>
<p style="margin-bottom: 0in"><a name="Current version">Current v<font size="3">ersio</font><font size="3"><span style="font-weight: normal">n:</a>
iRegulon Beta v0.10.3</span></font></p>
<p style="margin-bottom: 0in"><font size="3"><span style="font-weight: normal"><a name="Contact">Contact:</a>
	Bram Van de Sande 		</span></font><a href="mailto:Bram.VandeSande@med.kuleuven.be"><font size="3"><span style="font-weight: normal">Bram.VandeSande@med.kuleuven.be</span></font></a></p>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt"><font size="3"><span style="font-weight: normal">		Stein
Aerts 			</span></font><a href="mailto:Stein.Aerts@med.kuleuven.be"><font size="3"><span style="font-weight: normal">Stein.Aerts@med.kuleuven.be</span></font></a></font></p>
<p style="margin-bottom: 0in; font-weight: normal"></p>
<h2 class="western"><font  style="font-size: 11pt"><b>Contact</b></font></h2>
<p style="margin-bottom: 0in"><font size="3"><span style="font-weight: normal">Bram
Van de Sande 		</span></font><a href="mailto:Bram.VandeSande@med.kuleuven.be"><font size="3"><span style="font-weight: normal">Bram.VandeSande@med.kuleuven.be</span></font></a></p>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt"><font size="3"><span style="font-weight: normal">Stein
Aerts 				</span></font><a href="mailto:Stein.Aerts@med.kuleuven.be"><font size="3"><span style="font-weight: normal">Stein.Aerts@med.kuleuven.be</span></font></a></font></p>
<h2 class="western"><font  style="font-size: 11pt"><a name="Case study">Case study</a> </font>
</h2>
<h3 class="western"><a name="Gene based">gene based</a></h3>
<h4 class="western"><a name="1.3.1.1.input|outline"></a><a name="Gene based Input">input</a></h4>
<p style="margin-bottom: 0in"><br>
</p>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt">Open
the gene set as a network in Cytoscape (see Cytoscape tutorial). The
plugin can be started on 2 different ways: </font>
</p>
<ul>
	<li><p style="margin-bottom: 0in"><font  style="font-size: 11pt">start
	the panel (which gives the user a panel at the left),</font></p>
	</li><li><p style="margin-bottom: 0in"><font  style="font-size: 11pt">classical
	(which gives the user a window).</font></p>
</li></ul>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt">The
2 ways are the same for the classical analysis (other options of
iRegulon aren't ready yet).</font></p>
<p style="margin-bottom: 0in"><br>
</p>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt">For
the panel go to Plugins &gt; iRegulon &gt; Get the Panel.</font></p>
<p style="margin-bottom: 0in"><font  style="font-size: 11pt">For
the frame go to Plugins &gt; iRegulon &gt; Classical.</font></p>
<p style="margin-bottom: 0in"><br>
</p>
<p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt">Both
the panel as the frame has the same parameters:</font></p>
<ul>
	<li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">name<br></span></font><font  style="font-size: 11pt"><span lang="en-US">This
	is the name of the job. It will be the name visible in the output.
	This name will also be the suggested name for saving the results.</span></font></p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">species
	and nomenclature<br></span></font><font  style="font-size: 11pt"><span lang="en-US">This
	supply some information about the selected nodes.</span></font></p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">database
	system<br></span></font><font  style="font-size: 11pt"><span lang="en-US">This
	is the type of database (gene based or region based)</span></font></p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">database<br></span></font><font  style="font-size: 11pt"><span lang="en-US">This
	is the database (There are multiple possibilities).</span></font></p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">Overlap<br></span></font><font  style="font-size: 11pt"><span lang="en-US">This
	is only available with region based databases. The overlap stands
	for a percentage. This percentage is the amount of the </span></font><font  style="font-size: 11pt"><span lang="en-US">TF</span></font><font  style="font-size: 11pt"><span lang="en-US">
	that must be part of the defined region for the </span></font><font  style="font-size: 11pt"><span lang="en-US">TF</span></font><font  style="font-size: 11pt"><span lang="en-US">
	to be added to the set, if the </span></font><font  style="font-size: 11pt"><span lang="en-US">TF</span></font><font  style="font-size: 11pt"><span lang="en-US">
	lays on the border of the region.</span></font></p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">Delineation<br></span></font><font  style="font-size: 11pt"><span lang="en-US">A
	regular used region that is predefined. </span></font>
	</p>
	</li><li><p style="margin-bottom: 0in; line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">upstream
	and downstream<br></span></font><font  style="font-size: 11pt"><span lang="en-US">The
	base-pairs upstream and downstream the transcription start site
	(TSS). This lets the user choose </span></font><font  style="font-size: 11pt"><span lang="en-US">his/her
	preferred region.</span></font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">The
	score threshold (NEScore)<br>This is the minimal score threshold to
	consider a motif as being relevant. The best option for these score
	threshold is to choose between 2 or 2.5.</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">The
	ROC threshold for AUC calculation<br>The AUC value is the cut-off
	which to calculate the AUC. This value is used to compare and rank
	the motifs. It's a percentage, the cut-off of the ROC on the x-axis.
	1 means all genes on the x-axis.</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Genomic
	Threshold for Visualization<br>This is the x-axis cutoff for
	visualization of the ROC curve. This value corresponds with the top
	genes shown on the results.</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Minimal
	orthologous<br>This is how orthologous the TF may be. 0 shows all
	orthologous TFs, while 1 gives TFs of the same species as the gene
	set.</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Maximum
	motif similarity<br>This is the percentage of difference of the
	sequence of the TF and the sequence of the motif. 0 gives only exact
	or very similar matches.</font></p>
</li></ul>
<!--
<p style="line-height: 100%"><br><br>
</p>
<p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics13" src="iRegulon-tutorial_html_557c74b7.png"><br clear="LEFT"><font size="3"><i>Illustration
	1: The input panel of iRegulon</i></font></p>
<br><br>
<p></p>
-->
<p style="line-height: 100%"><br><br>
</p>
<p><img align="LEFT" border="0" name="graphics13" src="iRegulon-tutorial_html_557c74b7.png"><br clear="LEFT"><font size="3"><i>Illustration
	1: The input panel of iRegulon</i></font></p>
<br><br>
<p></p>

<p style="line-height: 100%; page-break-before: always"><a name="__DdeLink__719_74438011"></a>
<font  style="font-size: 11pt"><span lang="en-US">To continue
the analysis we use get the panel. All nodes are selected (by Ctrl +
a). As name </span></font><font  style="font-size: 11pt"><span lang="en-US">YY1_XiRen</span></font><font  style="font-size: 11pt"><span lang="en-US">
is suggested. Because multiple analysis with the same set will be
done, the name is changed to </span></font><font  style="font-size: 11pt"><span lang="en-US">YY1_XiRen</span></font><font  style="font-size: 11pt"><span lang="en-US">_</span></font><font  style="font-size: 11pt"><span lang="en-US">g</span></font><font  style="font-size: 11pt"><span lang="en-US">20kb_7species</span></font><font  style="font-size: 11pt"><span lang="en-US">
</span></font><font  style="font-size: 11pt"><span lang="en-US">(YY1_XiRen
as the set name, g20kb as gene based 20kb around TSS, 7 species)</span></font><font  style="font-size: 11pt"><span lang="en-US">.
 </span></font>
</p>
<p style="margin-left: 0.49in; line-height: 100%"><a name="line-62"></a><a name="line-63"></a>
<font  style="font-size: 11pt"><span lang="en-US">The set is a
</span></font><font  style="font-size: 11pt"><span lang="en-US">set
of Human genes, but not with HGNC symbols. Another tool available in
Cytoscape is used. CyThesaurus [</span></font><em><a href="http://www.ncbi.nlm.nih.gov/pubmed/20047655">The
BridgeDb framework: standardized access to gene, protein and
metabolite identifier mapping services</a></em><em>.</em> van Iersel
MP, Pico AR, Kelder T, Gao J, Ho I, Hanspers K, Conklin BR, Evelo CT.
BMC Bioinformatics. 2010 Jan 4;<strong>11</strong>(1):5. <font  style="font-size: 11pt"><span lang="en-US">]
is a tool for converting the names of nodes to other names. For more
information about the use of CyThesaurus, see there tutorial. Here
the BioMart database is used to convert the Ensemble gene ids to HGNC
symbols. By this conversion some genes are lost (6</span></font><font  style="font-size: 11pt"><span lang="en-US">85</span></font><font  style="font-size: 11pt"><span lang="en-US">
genes remains).</span></font></p>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">The
set is a set of Human genes, with HGNC symbols. For this analysis the
suggested gene based 20kb around TSS, 7 species will be used. The
standard parameters are used. For the gene name attributes an other
attribute is selected (this with the HGNC gene symbols, and not
canonical name (these are ensemble genes)). The amount of valid nodes
to submit is 6</span></font><font  style="font-size: 11pt"><span lang="en-US">85</span></font><font  style="font-size: 11pt"><span lang="en-US">.
Press submit to send the job to the iRegulon server.</span></font></p>
<p style="line-height: 100%"><br><br>
</p>
<h4 style="page-break-before: always" class="western"><font  style="font-size: 11pt"><a name="Gene based Running the analysis">Running
the analysis</a></font></h4>
<p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics2" src="iRegulon-tutorial_html_607f4df6.png"><br clear="LEFT"><font size="3"><i>Illustration
	2: Sending the request to the iRegulon server</i></font></p>
<br><br>
<p></p>
<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics3" src="iRegulon-tutorial_html_m5c5a28e2.png"><br clear="LEFT"><font size="3"><i>Illustration
	3: Waiting on the results of iRegulon</i></font></p>
<br><br>
<p></p>

<h4 style="page-break-before: always" class="western"><font  style="font-size: 11pt"><a name="Gene based Interpret the results">Interpret
the results</a></font></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">The
result of the analysis is a new Results panel (on the right of the
Cytoscape screen). This panel is separated into 2. Above the list of
motifs are visible, below the details of a selected motif in that
list. Because with this set it is known that the TF is YY1, YY1 is
typed in the search box. The top 10 motifs are motifs of YY1. </span></font><font  style="font-size: 11pt"><span lang="en-US">All
motifs with YY1 as transcription factor have a green check in the
first column (the others a red cross).</span></font></p>
<p style="line-height: 100%"><br><br>
</p>
<p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics4" src="iRegulon-tutorial_html_20096e35.png"><br clear="LEFT"><font size="3"><i>Illustration
	4: Results of the iRegulon analysis: above the motif panel, there
	under the detail of the selected motif.</i></font></p>
























<p style="line-height: 100%"><br><br>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">The
background of the motifs is the color of there cluster. The top 10
motifs are part of 3 different clusters. Important to know is that
these clusters are made by cluster-buster. These clusters are made on
sequence of the motif (not with TFs or TGs). Clusters can be handy
when no motif similarity is used. But here motif similarity is used
and </span></font><font  style="font-size: 11pt"><span lang="en-US">standing
on </span></font><font  style="font-size: 11pt"><span lang="en-US">the
TF YY1 of the motif with rank </span></font><font  style="font-size: 11pt"><span lang="en-US">1</span></font><font  style="font-size: 11pt"><span lang="en-US">
(flyfactorsurvey-pho_</span></font><font  style="font-size: 11pt"><span lang="en-US">SOLEXA</span></font><font  style="font-size: 11pt"><span lang="en-US">_</span></font><font  style="font-size: 11pt"><span lang="en-US">5</span></font><font  style="font-size: 11pt"><span lang="en-US">_FBgn0002521),
the motif of the TF is </span></font><font  style="font-size: 11pt"><span lang="en-US">actually</span></font><font  style="font-size: 11pt"><span lang="en-US">
this of the motif with rank </span></font><font  style="font-size: 11pt"><span lang="en-US">17</span></font><font  style="font-size: 11pt"><span lang="en-US">
(transfac_pro-M01744) </span></font><font  style="font-size: 11pt"><span lang="en-US">(seen
in the tooltip)</span></font><font  style="font-size: 11pt"><span lang="en-US">.
The </span></font><font  style="font-size: 11pt"><span lang="en-US">motif
with rank </span></font><font  style="font-size: 11pt"><span lang="en-US">17</span></font><font  style="font-size: 11pt"><span lang="en-US">
is in another cluster than motif </span></font><font  style="font-size: 11pt"><span lang="en-US">1</span></font><font  style="font-size: 11pt"><span lang="en-US">.
So these clusters only gives a indication. </span></font><font  style="font-size: 11pt"><span lang="en-US">I</span></font><font  style="font-size: 11pt"><span lang="en-US">f
motif similarity is used, use that to confirm cluster differences. </span></font>
</p>
<p style="line-height: 100%"><br><br>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
table has multiple columns. The first with the checks and crosses
indicates the search function. The second column has the rank of the
motif. The rank is derived by sorting the motifs by the NEScore. The
Enriched motif ID is the official name of the motif. The NES is the
normalized enrichment score. This score is an indication of how good
this motif is in this gene-set. The AUC is the area under the curve.
This is the area under the receiver-operating curve. This value is
used to calculate the NES. The clusternumber shows the same thing as
the background color of the motifs. The last 2 columns are the amount
of target genes and transcription factors.</font></p>
<p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics5" src="iRegulon-tutorial_html_4e298f82.png"><br clear="LEFT"><font size="3"><i>Illustration
	5: Result panel with the details of the selected motif</i></font></p>
<br><br>
<p></p>



























<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">A
detail frame can be obtained by selecting a motif, for example motif
</span></font><font  style="font-size: 11pt"><span lang="en-US">with
rank 1</span></font><font  style="font-size: 11pt"><span lang="en-US">
and selecting the TF YY1. Right-clicking on this TF will open a new
window with the </span></font><font  style="font-size: 11pt"><span lang="en-US">details.
Here in this frame it's possible to compare the motifs for the
similarity. (In this case the motifs are similar, but reverse
complement). </span></font><font  style="font-size: 11pt"><span lang="en-US">In
this frame the input parameters are displayed above. There under the
motif with all the details. And under the motif the TF is displayed,
with the TGs.</span></font></p>
<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics6" src="iRegulon-tutorial_html_7f63af4f.png"><br clear="LEFT"><font size="3"><i>Illustration
	6: The details of Motif 1 and the transcription factor YY1</i></font></p>
<br><br>
<p></p>

<h4 style="page-break-before: always" class="western"><a name="1.3.1.4.Visualization|outline"></a>
<font  style="font-size: 11pt"><a name="Gene based Visualization">Visualization</a></font></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">The
important part of Cytoscape is the visualization of the interactions
between genes, proteins, TFs, etc. This can also been done with
iRegulon. </span></font><font  style="font-size: 11pt"><span lang="en-US">When
a motif is selected, the second row become enabled (when this motif
has one or more TFs). </span></font><font  style="font-size: 11pt"><span lang="en-US">First
the right TF must be selected. This is possible by selecting it in
the combobox. If the wanted TF isn't available for that motif, it's
possible to type it in this combobox. </span></font>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">There
are a few options to visualize a network of this TF and his TGs. The
2 most used are available as a button in front of the combobox. All
draw options are available when right-clicking on the motif.</font></p>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
possible draw options are:</font></p>
<ul>
	<li><p style="line-height: 100%"><font  style="font-size: 11pt">draw
	TF, TGs and there interactions on the current network,</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">create
	a new network with the selected TF, his TGs and there interactions,</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">draw
	the interactions on the current network,</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">draw
	a new network with combined edges (from iRegulon).</font></p>
</li></ul>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">The
first draw action (draw TF, TGs and there interactions on the current
network) will add new nodes if the TF or TGs don't already exists. </span></font><font  style="font-size: 11pt"><span lang="en-US">If
the TF or TGs already exists there name is colored red in the lists.
It may appear that only one node is added to the current network.
This is because there is no layout added to the view. So the nodes
are structured on a special way, the new nodes can be structured as
that manualy, or with a, by Cytoscape provided, layout. (Layout in
the Cytoscape menu). The node that is the TF has as attribute
Regulatory Function the value Regulator. Also all the drawn motifs of
that node are summed in the attribute Motif. </span></font><font  style="font-size: 11pt"><span lang="en-US">The
edges has as attributes the motif, Regulatory Gene (the TF), Target
Gene (the TG), Regulatory function (predicted) and as interaction
regulates and the motif name.</span></font></p>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">A
new network can be created by the create a new network action. In
this action a new network is created. The only nodes in this network
are the TF and his TGs. These nodes have the same attributes as by
the draw action. This is also the case for the edges. The nodes have
the default grid layout of Cytoscape (Layout &gt; Cytoscape Layouts &gt;
Grid Layout). </span></font><font  style="font-size: 11pt"><span lang="en-US">The
new network also has an other view (VizMapper). The TF is a round,
green node. </span></font><font  style="font-size: 11pt"><span lang="en-US">The
TGs are white, square nodes. The </span></font><font  style="font-size: 11pt"><span lang="en-US">edges
are directed from the TF to the TG. Edges are grouped. The edges with
the same motif will have the same color. </span></font>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
possibility to just draw the interactions on a network also exists.
This will only draw the edges. Edges can only been drawn when the
selected TF exists in that network. When no TF is available, no edges
will be drawn.</font></p>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
combined edges will copy the network to a new network (nodes, edges
and layout). All edges drawn by iRegulon will be merged. This means
when there are multiple edges between a TF and a TG, these edges will
be replaced by one edge. This edge will keep all the information of
the different edges. (but some functionality as linkout (only of
region based databases, see later) won't work any more).</font></p>

<p style="line-height: 100%"><br><br>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">A
new network is created of the TF YY1 and his TGs for motif 1. YY1 is
drawn as a green circle. The target genes are white squares. The
layout is the regular grid layout of cytoscape.</font></p>

<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics7" src="iRegulon-tutorial_html_m579a1a08.png"><br clear="LEFT"><font size="3"><i>Illustration
	7: The network of motif flyfactorsurvey-pho_SOLEXA_5_FBgn0002521
	with as transcription factor YY1 and all his target genes</i></font></p>






<p style="line-height: 100%"><br><br>
</p>
<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics1" src="iRegulon-tutorial_html_m2651faf6.png"><br clear="LEFT"><font size="3"><i>Illustration
	8:  Whole iRegulon in one picture: left the input panel, in the
	middle a network drawn of a motif, and right the output of iRegulon,
	with the drawn motif selected</i></font></p>
<br><br>
<p></p>

<h4 style="page-break-before: always" class="western"><a name="1.3.1.5.Save and Export|outline"></a>
<a name="Gene based Save and Export">Save and Export</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt">Saving
the results can by clicking the diskette. This will save the results
as a .irf. This file can be loaded by selecting Plugins &gt; iRegulon
&gt; Load. The output in the results panel will be saved (only that
selected run, and no networks). Created networks has to be saved on
the regular way in Cytoscape (File &gt; save). The results can also
be saved as a tab-delimited file. This by clicking on the symbol next
the diskette. The file has the following columns:</font></p>
<ul>
	<li><p style="line-height: 100%"><font  style="font-size: 11pt">rank</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">motif
	id</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">AUC</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">NES</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Cluster
	Code</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Transcription
	factor</font></p>
	</li><li><p style="line-height: 100%"><font  style="font-size: 11pt">Target
	genes</font></p>
</li></ul>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
columns are separated by a tab. The transcription factors and target
genes are separated from each other by a comma. The first row in the
file is a heading.</font></p>

<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics10" src="iRegulon-tutorial_html_2cfbf1fd.jpg"><br clear="LEFT"><font size="3"><i>Illustration
	9: An example of the tab-delimited file</i></font></p>
<br><br>
<p></p>

<h3 style="page-break-before: always" class="western"><a name="Region based">Region based</a></h3>
<h4 class="western"><a name="Region based Input">Input</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt">Region
based analysis are done in a similar way as <a href="#1.3.1.1.input|outline">gene
based analysis</a>. </font>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
input has more parameters available (overlap, delineation or upstraem
and downstream). First the database system must be changed to region
based. Automaticaly the screen is updated and the overlap and the
radio-buttons are enabled. A database can be chosen. The database
used in this tutorial is the All regulatory features database. In
this tutorial the gene based database was with 20kb around the TSS
and 7 species. So the delineation 20kb around the TSS is chosen. The
overlap is standard 0.4 and is not changed. Because delineation is
used, no base-pairs upstream and downstream are needed (this is only
when no delineation is used). The name is changed to
YY1_XiRen_r20kb_all (of YY1_XiRen as the set, r20kb as region based
20kb around TSS, all as all regulatory features). </font>
</p>

<h4 class="western"><a name="Region based Running the analysis">Running the analysis</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt">The
calculations with region based databases require a bit more time than
gene based databases. The output of the region based analysis is the
same as of the gene based. The same options are available. </font>
</p>

<h4 style="page-break-before: always" class="western"><a name="Region based Interpret the results">Interpret the
results</a></h4>
<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics8" src="iRegulon-tutorial_html_43caaac.png"><br clear="LEFT"><font size="3"><i>Illustration
	10: Region based input and output</i></font></p>
<br><br>
<p></p>

<p style="line-height: 100%"><font  style="font-size: 11pt">The
results gives slightly different results. So it's recommended to do
multiple analysis if the TF isn't known. Here is YY1 one of the top
found TFs (actually are the other found TFs biological related to
YY1). The available actions are the same as by gene based databases
(so see gene based databases). </font>
</p>
<p style="line-height: 100%"><font  style="font-size: 11pt">An
extra option comes with region based databases. This option is the
link-out. First a new network is drawn of the first motif
(flyfactorsurvey-phol_SANGER_5_FBgn0035997), with as TF YY1 (by
selecting the first motif, select the transcription factor YY1 and
click on the draw network button).</font></p>
<h4 class="western"><br><br>
</h4>
<h4 style="page-break-before: always" class="western"><a name="Region based Visualization">Visualization</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt">This
is done on the same way as for <a href="#1.3.1.4.Visualization|outline">gene
based databases</a>.</font></p>
<h4 class="western"><a name="Region based Save and Export">Save and Export</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">This
is done on the same way as for </span></font><a href="#1.3.1.5.Save and Export|outline">gene
based databases</a><font  style="font-size: 11pt"><span lang="en-US">.</span></font></p>
<h4 class="western"><a name="Region based Link Out">Link Out</a></h4>
<p style="line-height: 100%"><font  style="font-size: 11pt"><span lang="en-US">Linkout
is possible by selecting an edge. This edge must be an edge drawn by
iRegulon. Right-clicking on the edge and choose linkout &gt;
iRegulon. Your default web-browser will open. </span></font><font  style="font-size: 11pt"><span lang="en-US">The
location is showed as a track in the UCSC website. The track has a
unique name. This name is the motif name underscore target gene name
underscore transcription factor name (</span></font><font color="#000000"><font  style="font-size: 11pt"><span lang="en-US"><span style="font-style: normal"><span style="text-decoration: none"><span style="font-weight: normal">here
</span></span></span></span></font></font><a href="http://genome.ucsc.edu/cgi-bin/hgTrackUi?hgsid=268615413&amp;c=chr3&amp;g=ct_iRegulonflyfactorsurveypholSANGER5FBgn0035997TRA2BYY1_2598"><span style="font-variant: normal"><font color="#000000"><font face="Times New Roman"><font  style="font-size: 11pt"><span style="font-style: normal"><span style="text-decoration: none"><span style="font-weight: normal">iRegulon_flyfactorsurvey-phol_SANGER_5_FBgn0035997_TRA2B_YY1</span></span></span></font></font></font></span></a><span style="font-variant: normal"><font color="#000000"><font face="Times New Roman"><font  style="font-size: 11pt"><span style="font-style: normal"><span style="text-decoration: none"><span style="font-weight: normal">).
The track has a header: iRegulon enhancer location for transcription
factor (and his name) and target gene (and his name). In this case is
it: iRegulon enhancer location for transcription factor YY1 and
target gene TRA2B. If 10x is pressed in the browser, the tracks are
zoomed out. Here it is possible to see the exact location of the
motif in the genome, in perspective of the target gene.</span></span></span></font></font></font></span></p>
<p style="font-variant: normal; font-style: normal; font-weight: normal; line-height: 100%; text-decoration: none">
</p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics11" src="iRegulon-tutorial_html_m35243a8c.jpg"><br clear="LEFT"><font size="3"><i>Illustration
	11: The transcription factor binding site (TFBS) visible in the UCSC
	browser, after zoom out 10x. The TFBS is located in the first intron
	after the transcription start side (TSS).</i></font></p>
<br><br>
<p></p>



<p style="line-height: 100%"></p><p style="margin-top: 0.08in"><img align="LEFT" border="0" name="graphics12" src="iRegulon-tutorial_html_11f6acd5.jpg"><br clear="LEFT"><font size="3"><i>Illustration
	12: The transcription factor binding site (TFBS) visible in the UCSC
	browser, after zoom out 10x. The TFBS is located in the first intron
	after the transcription start side (TSS).</i></font></p>
<br><br>
<p></p>






<p style="line-height: 100%"><br><br>
</p>
<div type="FOOTER">
	<p style="margin-top: 0.2in; margin-bottom: 0in">	<font face="Times New Roman, serif">&copy;</font><font face="Times New Roman, serif">
	2012 KU Leuven</font></p>
</div>

</body></html>
